### PR TITLE
LICENSE: split license file in standard BSD 3-clause and bundled.

### PR DESCRIPTION
**Pull request by rgommers**
Reason: the GitHub license detection method relies on the LICENSE
file matching for at least 95% with a standard license.
Adding even one more sentence changes the license displayed in the
GitHub UI and also in the GitHub API that can be queried from
BSD to "other".

See https://github.com/numpy/numpy/issues/13447 for more details.

Note that this split license is what GitHub recommends to do,
and is also what we do for wheels, where we append to the license
file at build time.

### CODE REVIEW

**Comment by rgommers** [comment]
Note: I will update the numpy-wheels repo after this is merged.

**Comment by rgommers** [comment]
This is green. Would be nice if someone can verify this locally and merge, since I'd like to get this resolved so Tidelift dashboard is happy.

File: setup.py [hunk 1]
```diff
@@ -186,12 +186,45 @@ def check_submodules():
...
+        with open(self.f1, 'r') as f1:
+            self.bsd_text = f1.read()
+
+        with open(self.f1, 'a') as f1:
+            with open(self.f2, 'r') as f2:
+                self.bundled_text = f2.read()
+                f1.write('\n\n')
+                f1.write(self.bundled_text)
+
+        os.remove(self.f2)
```
Comment (charris) [inline setup.py]: Could just prune this in `MANIFEST.in`?

**Comment by charris** [comment]
How about making a `licenses` subdirectory that contains all the licenses used for wheels as well as a `LICENSE_github.txt` and `LICENSE_complete.txt`. The setup.py would then copy the appropriate license for source files/wheels to the root `LICENSE.txt`. That would make maintaining the wheels licenses easier as well as minimizing  the damage in case of failure as all can be fixed with a `git reset --hard HEAD`.

**Comment by charris** [comment]
Ideally this would be fixed on the github end.

**Comment by charris** [comment]
I'm a bit concerned about setup.py running in the wheels build, won't that modify the text?

**Comment by charris** [comment]
Is `LICENSE.md` also a valid license file?

**Comment by shoyer** [comment]
> Note that this split license is what GitHub recommends to do

could you kindly point me to where you found this guidance?

> I'd like to get this resolved so Tidelift dashboard is happy.

Does Tidelift really care about this? Apparently the license detection program has a bug specifically around the Apache license (see my comment in #13447), but note that the Apache Software Foundation's guidelines  that explicitly tell you to include mention of bundled licenses in the LICENSE file: 
http://www.apache.org/dev/licensing-howto.html#permissive-deps

**Comment by rgommers** [comment]
> could you kindly point me to where you found this guidance?

From the description of gh-13447: https://help.github.com/en/articles/licensing-a-repository#detecting-a-license

> Does Tidelift really care about this? 

What they do is use both the libraries.io API and the GitHub API to get license information. If there's a mismatch, then it gets flagged as an issue needing to be resolved. This is a sensible approach to verifying license info at scale. 

In our case, libraries.io correctly says BSD-3 while GitHub says "other".

**Comment by rgommers** [comment]
> Is LICENSE.md also a valid license file?

not sure. I prefer `LICENSE.txt` or `LICENSE`.

> I'm a bit concerned about setup.py running in the wheels build, won't that modify the text?

That's why I stated that I would fix the `numpy-wheels`  repo after.

> Ideally this would be fixed on the github end.

Me too, but I don't have much energy/time for that at the moment. This is by far the easier route.

> How about making a licenses subdirectory that contains all the licenses used for wheels as well as a LICENSE_github.txt and LICENSE_complete.txt. The setup.py would then copy the appropriate license for source files/wheels to the root LICENSE.txt. That would make maintaining the wheels licenses easier as well as minimizing the damage in case of failure as all can be fixed with a git reset --hard HEAD.

This is another option. However it's equally complicated. We then include licenses in this repo for code that's not actually in this repo. We then need to strip it out of the sdists. And, importantly, the `zip` or `tar.gz` file that then appears on https://github.com/numpy/numpy/releases and with `Clone and Download` on the main page of the repo will include those licenses. That's a big downside, I'd much prefer not to have anything that says LGPL in here.

**Comment by rgommers** [comment]
> That would make maintaining the wheels licenses easier 

On second thought, I think it's important to treat `numpy-wheels` the same as we treat other redistributors of NumPy. It just happens to be that we are the redistributor in that case. But there can be others redistributing wheels, as well as conda packages, Linux and macOS package managers (which often build against ATLAS), etc. Those all do not have any interactions with the `numpy-wheels` code, and we shouldn't bother those packagers with licenses for code that's not in this repo.

> as well as minimizing the damage in case of failure as all can be fixed with a `git reset --hard HEAD`.

there's very little chance of that with this context manager. also, this is only triggered for releases, never for regular install or development. so doesn't seem much of an issue to me

**Comment by shoyer** [comment]
> What they do is use both the libraries.io API and the GitHub API to get license information. If there's a mismatch, then it gets flagged as an issue needing to be resolved. This is a sensible approach to verifying license info at scale.

If we're doing this for compliance purposes per Tidelift's request, it seems like we should actually ask them what they prefer rather than just making an assumption.

Perhaps we really should call NumPy's license "Other" given that it has bundled code with other licenses, I don't know!

**Comment by rgommers** [comment]
> If we're doing this for compliance purposes per Tidelift's request,

I don't see it only in that light. I think this pointed out an issue that we weren't aware of - I wasn't even aware GitHub offered an API to query the license. And if you'd asked me that before last week, I'd also have said that of course it should say NumPy is BSD-3.

> it seems like we should actually ask them what they prefer rather than just making an assumption.

I don't think they will answer. Anyway, it's up to us what we want to say. All they're asking is to fix an inconsistency between two sources of information.

> Perhaps we really should call NumPy's license "Other" given that it has bundled code with other licenses, I don't know!

I don't think so, that would be very weird. We've always explicitly communicated that our license is BSD-3, and only incorporate code that is BSD-3 or (in some exceptional case) not BSD-3 but compatible with it.

An example: on https://pypi.org/project/numpy/ we say "All NumPy wheels distributed on PyPI are BSD licensed."

In our [setup.py](https://github.com/numpy/numpy/blob/master/setup.py#L372) we say "BSD". 

Also note that in the past we've gotten explicit legal advice about the way we're shipping wheels, and whether bundling gfortran/openblas etc. is fine (it is).

**Comment by shoyer** [comment]
> > it seems like we should actually ask them what they prefer rather than just making an assumption.
> 
> I don't think they will answer

Really? They're paying us $$$$ on behalf of their customers so I would be rather shocked if they don't bother to answer our emails :).

Anyways, I'm happy to leave this up to your judgment.

**Comment by rgommers** [comment]
Of course I meant: they will answer my email, but they will say it's our decision.

File: setup.py (hunk 1 above)
Comment (rgommers) [reply setup.py]: Good suggestion, done. That also made me spot another issue: README.md was not included in the sdist - fixed that as well.

**Comment by rgommers** [comment]
> Anyways, I'm happy to leave this up to your judgment.

Thanks Stephan.

Okay, this is kind of the minimal fix possible, and I'm planning to merge it tomorrow unless there are more comments.

**Comment by charris** [comment]
Thanks Ralf.

**Comment by rgommers** [comment]
Thanks Chuck, Stephan.

Annoyingly, this still seems to have an issue. Need to rename `LICENSE_bundled.txt` to `LICENSES_bundled.txt`:( I'll do that now

**Comment by rgommers** [comment]
Opened an issue on the Licensee repo, this is getting ridiculous.

**Pull request merged**